From be94b3ecae3600e5b9e710d71347a7b4340e96fc Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Wed, 23 Jul 2014 13:46:20 -0700
Subject: [PATCH 1/1] power: dc_xpwr_charger: check USB host mode at boot

When the device boots with OTG cable connected, the charger driver
does not get notification from the extcon-3gpio driver. This causes
the charger to start charging as VBUS is already present.

This patch checks the host mode state through extcon driver. So that
it will not commence charging if in USB host mode.

Issue: GMIN-2695
Change-Id: Ibd4cd9b8f35803218b89f4083a472de897bc8509
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 drivers/power/dc_xpwr_charger.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/power/dc_xpwr_charger.c b/drivers/power/dc_xpwr_charger.c
index 8d7f185..bcf3b68 100644
--- a/drivers/power/dc_xpwr_charger.c
+++ b/drivers/power/dc_xpwr_charger.c
@@ -796,13 +796,18 @@ static void dc_xpwr_otg_event_worker(struct work_struct *work)
 		gpio_direction_output(info->pdata->otg_gpio, info->id_short);
 }
 
+static bool is_usb_host_mode(struct extcon_dev *evdev)
+{
+	return !!evdev->state;
+}
+
 static int dc_xpwr_handle_otg_event(struct notifier_block *nb,
 				   unsigned long event, void *param)
 {
 	struct pmic_chrg_info *info =
 	    container_of(nb, struct pmic_chrg_info, id_nb);
 	struct extcon_dev *edev = param;
-	int usb_host = !!edev->state;
+	int usb_host = is_usb_host_mode(edev);
 
 	dev_info(&info->pdev->dev,
 		"[extcon notification] evt:USB-Host val:%s\n",
@@ -897,6 +902,9 @@ static int pmic_chrg_probe(struct platform_device *pdev)
 	if (ret)
 		dev_err(&pdev->dev, "failed to register extcon notifier\n");
 
+	if (info->cable_obj.edev)
+		info->id_short = is_usb_host_mode(info->cable_obj.edev);
+
 	info->psy_usb.name = DEV_NAME;
 	info->psy_usb.type = POWER_SUPPLY_TYPE_USB;
 	info->psy_usb.properties = pmic_chrg_usb_props;
-- 
1.8.3.2

